### PR TITLE
feat(http): Support `maxRetryAfter` option

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Pass `maxRetryAfter` to `got` via `hostRules`.

## Context

- Ref: #25499

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — DISMISSED** [review]

File: lib/util/http/index.ts [hunk 1]
```diff
@@ -132,6 +132,11 @@ export class Http<Opts extends HttpOptions = HttpOptions> {
   ) {
     this.options = merge<GotOptions>(options, { context: { hostType } });
 
+    this.options.retry = {
+      limit: 2,
+      maxRetryAfter: 60,
+    };
```
Comment (viceice) [inline lib/util/http/index.ts:138]: why not adding to the merge call above?

**Comment by zharinov** [comment]
- We could merge this one first, then continue with #25859 if necessary

**Review by viceice — APPROVED** [review]

**Review by rarkins — COMMENTED** [review]
Does this affect how timeouts work for non-retry scenarios?

And how does it work if the maxRetryAfter is e.g. 120 while the default timeout is 60?

**Comment by zharinov** [comment]
> Does this affect how timeouts work for non-retry scenarios?

By default, `maxRetryAfter` is inherited from `timeout`. Here we decouple one from another, so now they're completely separated I guess.

> And how does it work if the maxRetryAfter is e.g. 120 while the default timeout is 60?

If `120` is set in `hostRules`, it should override `60`. If nothing was set in `hostRules`, it will be 60 seconds by default.

**Comment by zharinov** [comment]
Just tested it, `timeout` is not being implicitly increased in the `got` side, so starting from `61`, requests are immediately cancelled without any delays and retries. Fortunately, we could set `options.timeout = Math.max(options.timeout, options.retry.maxRetryAfter)`

**Comment by rarkins** [comment]
Yes but here's the thing:
- I want real timeouts to "fail fast" and not wait minutes for a server which won't respond. I don't want this time increased just so we can support retry-after
- If we get a real response with retry-after header then I consider that different from a timeout and want to control it via maxRetryAfter

So therefore increasing timeout to accommodate maxRetryAfter is not good for the majority of timeouts as now they will just get higher. I think perhaps we need more granularity than "timeout", and maybe got offers that?

**Comment by viceice** [comment]
yes, got offers more granular timeouts.

this PR doesn't increase the current timeout. got timeouts are currently buggy when using auto retry. so even we set a got timeout to 60s, the real timeout can be 180s because of the retries and some internal got inconsistencies in current version. later versions of got fixed that bug they are esm only.

maybe we should switch to another library like `undici` which doesn't has the issues and isn't esm only. `undici` is also the implementation for the node buildin fetch.

**Comment by zharinov** [comment]
Maybe it's not necessary to mock the whole network stack anyway, but it's another discussion

**Pull request closed without merging**